rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,394 @@
+the identifiers must be deleted when the final object is created. The
+interface is a simplified form of what is used in the `file-source`
+library with the unused bits removed.
+
+```rust
+pub struct Checkpointer<K, V> {
+    tmp_file_path: PathBuf,
+    stable_file_path: PathBuf,
+    checkpoints: Arc<CheckpointsView>,
+}
+
+pub struct CheckpointsView {
+    checkpoints: DashMap<MultipartUploadKey>,
+}
+
+type MultipartUploadKey = String;
+
+/// Multipart upload state data
+struct MultipartUploadData {
+    /// the identifier for the upload process
+    upload_id: String,
+    /// A list of uploaded parts e_tags. The part number is derived from the index.
+    parts: Vec<String>,
+}
+```
+
+## Drawbacks
+
+- By increasing the time over which events are written to the sink, we
+  are increasing the chances of data loss due to crashes and network
+  interruptions.
+
+## Prior Art
+
+The [S3 output plugin](https://docs.fluentd.org/output/s3) for fluentd
+creates one upload per hour by default. It uses a disk buffer to store
+the object before uploading.
+
+## Alternatives
+
+### Buffer the full large object locally
+
+Instead of using the multipart upload API, Vector could save the data
+that would compose the final object into a local disk buffer, and then
+upload it in the background. This has several disadvantages:
+
+1. It requires local storage for the complete object.
+2. It creates complications for handling upload failures, as
+   acknowledgements for that upload are disconnected from any source
+   data.
+
+### Upload parts as regular objects
+
+Vector could also upload the individual batches as regular objects and
+then go back and assemble the result into a larger object after all
+the uploads are complete. Much of the proposal described above would
+be the same, but the ordering would be changed and an additional step
+is required:
+
+1. Upload batches as regular objects, making note of object names.
+2. When the final object is ready, create it with
+   `CreateMultipartUpload`.
+3. Add the previous objects using `UploadPartCopy`.
+4. Assemble the final object with `CompleteMultipartUpload`.
+5. Delete the intermediate objects that were assembled into the final
+   object.
+
+The downside with this scheme is the additional complexity of tracking
+the intermediate object names and the deletion step at the end of the
+process, both of which are handled automatically when uploading parts
+as above.
+
+### Use the multipart upload copy to effect append behavior
+
+One of the actions possible when creating a multipart object is to
+copy data from an existing object. This can be used to both copy data
+from and replace an object at the same time, allowing for effective
+append (or prepend) behavior without using extra bandwidth. It is
+unclear, however, if this behavior will interact badly with object
+tracking sources like SQS, or if AWS would penalize this action
+pattern for high volume uploads.
+
+## Outstanding Questions
```

Comment:
The upload structure and parts still live in the bucket until they are either purged by the bucket lifecycle configuration or are assembled into an object, which can be done after the fact. They are just not visible as normal objects.

Reply:
Yeah, that's my concern. Most S3 buckets have a life cycle rule that'll delete stale multipart uploads, which would result in data loss. So, to me, this violates the guarantee that e2e ack'ing implies. The data isn't actually written. This makes me want to upload the parts as regular objects that get assembled later. This has the advantage of:

1. Making the data immediately available for analysis with tools like Athena.
2. Prevents the case above where the data gets deleted or lost.

To prevent data loss, how would someone practically reassembled orphaned parts? I've never seen a tool that does this since it's usually the responsibilities of the tool uploading the data.

Reply:
There are API calls to list current multipart uploads and the associated uploaded parts. The orphaned parts could then be reassembled from this data.

Another downside to uploading as regular parts is that the data would show up twice to anything that is monitoring the uploads in real time (since the events would show both in the individual files and in the assembled larger object).

Reply:
I see. Then I'd like to think about how Vector could solve this in the event of a Vector process crash.

Reply:
I should have clarified earlier: The multipart upload identifiers and uploaded part information will necessarily be written to a checkpoint (needed to handle non-crash reloads and restarts as well) so one of the startup tasks will have to be to check if any of those uploads are past their completion time and assemble them.

Reply:
Generally I think it's a good practice to assume at startup that we may have crashed in the middle of an operation and do these kinds of checks to try to rebuild our state as much as possible. It might even be worth investing some time in a subcommand or similar that can inspect this state on disk without needing to start up Vector with a matching config.